File: src/main/python/genuvem.py
from pyspark.sql import SparkSession
import uuid
import os
import sys
import argparse

parser = argparse.ArgumentParser(prog="Genuvem")
parser.add_argument("--databank", "-d", type=str, default="sars-cov-2-2021")
parser.add_argument("--query", "-q", type=str, default="sars-cov-1.fasta")
args = parser.parse_args(sys.argv[1:])

databank = args.databank
query_path = "/app/genoogle/queries/exp2/" + args.query
script_path = "/app/genoogle/run_genoogle.sh"

# Initialize Session
spark = SparkSession.builder.master("yarn").appName('Genuvem').getOrCreate()
sc = spark.sparkContext

# Get the number of available nodes
node_count = args.partitions

# Generate unique id and build the args array
run_id = str(uuid.uuid4())
run_args = [script_path, databank, run_id, args.query]
envs = {"GENOOGLE_HOME": os.getenv("GENOOGLE_HOME")}

# Set '>' as delimiter for FASTA files
conf = sc._jsc.hadoopConfiguration()
conf.set("textinputformat.record.delimiter", ">")

# Add '>' back to each sequence read from the query file
# and repartition the RDD by the number of available nodes
rdd = sc \
    .textFile(query_path) \
    .map(lambda x: x.strip()) \
    .filter(lambda x: x) \
    .map(lambda x: '>' + x) \
    .repartition(node_count)

# Run Genoogle and collect the HDFS path of the output files
output_files = rdd.pipe(" ".join(run_args), envs).collect()

print(f"Run {run_id} finished successfully. HDFS output files:")
print(output_files)

File: src/main/scala/Genuvem.scala
import org.apache.spark.sql.SparkSession
import java.util.UUID.randomUUID

object Genuvem {
  def main(args: Array[String]): Unit = {

    // Pipeline parameters
    val databank = args(0) // "sars-cov-2-2021"
    val queryFile =  args(1) // "/search/sars-cov-1-query.fasta"

    val queryPath = "file:///app/genoogle/queries/exp2/" + queryFile
    val scriptPath = "/app/genoogle/run_genoogle.sh"

    // Initialize Session
    val spark = SparkSession
      .builder
      .master("yarn")
      .appName("Genuvem")
      .getOrCreate

    val sc = spark.sparkContext
    val nodeCount = sc.getConf.getInt("spark.executor.instances", 1)
    println(s"nodeCount is $nodeCount")

    // Generate unique id and build the args array
    val runId = randomUUID.toString
    val runArgs = Seq(scriptPath, databank, runId, queryFile)
    val envs = Map("GENOOGLE_HOME" -> sys.env("GENOOGLE_HOME"))

    // Set '>' as delimiter for FASTA files
    val conf = sc.hadoopConfiguration
    conf.set("textinputformat.record.delimiter", ">")

    // Add '>' back to each sequence read from the query file
    // and repartition the RDD by the number of available nodes
    val rdd = sc.textFile(queryPath)
      .map(x => x.trim())
      .filter(x => x.nonEmpty)
      .map(x => '>' + x)
      .repartition(nodeCount)

    // Run Genoogle and collect the HDFS path of the output files
    val outputFiles = rdd.pipe(runArgs, envs).collect

    println(s"Run $runId finished successfully. HDFS output files:")
    outputFiles.map(println)
  }
}
